// weapp-tailwindcss/dist/index-De0cmOEK.js
'use strict';

var options = require('./options-BAFFNEBI.js');
var stream = require('node:stream');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var stream__default = /*#__PURE__*/_interopDefaultCompat(stream);

const Transform = stream__default["default"].Transform;
function createPlugins(options$1 = {}) {
    const opts = options.getOptions(options$1);
    const { templateHandler, styleHandler, patch, jsHandler, setMangleRuntimeSet, tailwindcssBasedir } = opts;
    let runtimeSet = new Set();
    patch === null || patch === void 0 ? void 0 : patch();
    const twPatcher = options.createTailwindcssPatcher();
    function transformWxss() {
        const transformStream = new Transform({ objectMode: true });
        transformStream._transform = function (file, encoding, callback) {
            return options.__awaiter(this, void 0, void 0, function* () {
                runtimeSet = twPatcher.getClassSet({
                    basedir: tailwindcssBasedir
                });
                setMangleRuntimeSet(runtimeSet);
                const error = null;
                if (file.contents) {
                    const code = yield styleHandler(file.contents.toString(), {
                        isMainChunk: true
                    });
                    file.contents = Buffer.from(code);
                }
                callback(error, file);
            });
        };
        return transformStream;
    }
    function transformJs() {
        const transformStream = new Transform({ objectMode: true });
        transformStream._transform = function (file, encoding, callback) {
            const error = null;
            if (file.contents) {
                const { code } = jsHandler(file.contents.toString(), runtimeSet);
                file.contents = Buffer.from(code);
            }
            callback(error, file);
        };
        return transformStream;
    }
    function transformWxml() {
        const transformStream = new Transform({ objectMode: true });
        transformStream._transform = function (file, encoding, callback) {
            const error = null;
            if (file.contents) {
                const code = templateHandler(file.contents.toString(), {
                    runtimeSet
                });
                file.contents = Buffer.from(code);
            }
            callback(error, file);
        };
        return transformStream;
    }
    return {
        transformWxss,
        transformWxml,
        transformJs
    };
}

exports.createPlugins = createPlugins;
