// weapp-tailwindcss/dist/replace.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var escape = require('@weapp-core/escape');
require('@ast-core/escape');
var shared = require('./shared-DtRIj-D_.js');

function replaceWxml(original, options = {
    keepEOL: false,
    escapeMap: escape.SimpleMappingChars2String
}) {
    const { keepEOL, escapeMap, mangleContext } = options;
    let res = original;
    if (!keepEOL) {
        res = res
            .replaceAll(/[\n\r]+/g, '');
    }
    if (mangleContext) {
        res = mangleContext.wxmlHandler(res);
    }
    res = escape.escape(res, {
        map: escapeMap
    });
    return res;
}

Object.defineProperty(exports, "MappingChars2String", {
    enumerable: true,
    get: function () { return escape.MappingChars2String; }
});
Object.defineProperty(exports, "MappingChars2StringEntries", {
    enumerable: true,
    get: function () { return escape.MappingChars2StringEntries; }
});
Object.defineProperty(exports, "SYMBOL_TABLE", {
    enumerable: true,
    get: function () { return escape.SYMBOL_TABLE; }
});
Object.defineProperty(exports, "SimpleMappingChars2String", {
    enumerable: true,
    get: function () { return escape.SimpleMappingChars2String; }
});
Object.defineProperty(exports, "SimpleMappingChars2StringEntries", {
    enumerable: true,
    get: function () { return escape.SimpleMappingChars2StringEntries; }
});
exports.replaceCss = shared.internalCssSelectorReplacer;
exports.replaceJs = replaceWxml;
